File: src/recall/journal.ts
/**
 * Journal writers / readers — `MemoryJournal` and `NdjsonFileJournal`.
 *
 * Both expose the same `Journal` interface so callers (engine
 * simulator, replay tooling, tests) pick storage at construction and
 * never branch on it after.
 *
 * The on-disk format is NDJSON: one entry per line, UTF-8, append-only.
 * No JSON-array wrapping — that would require rewriting the file's
 * closing `]` on every append, which defeats "append-only".
 */

import { appendFileSync, existsSync, readFileSync, statSync, writeFileSync, renameSync } from 'fs';
import type {
  Hex64,
  JournalManifest,
  SchemaVersion,
  SpinJournalEntry,
} from './types.js';
import { RECALL_SCHEMA_VERSION, ZERO_HASH } from './types.js';
import { sealEntry, sealManifest, verifyChain } from './integrity.js';

/**
 * Minimal storage interface. `head_hash` is the running chain tail —
 * callers do not pass it; the journal tracks it.
 */
export interface Journal {
  /** Append a finalized entry. The journal stamps `prev_hash` +
   * `entry_hash` for you given the entry minus those two fields. */
  append(draft: Omit<SpinJournalEntry, 'prev_hash' | 'entry_hash'>): SpinJournalEntry;
  /** Read all entries (memory journal). FileJournal also supports
   * streaming via `iterate` for large files. */
  readAll(): SpinJournalEntry[];
  /** Current chain head (last `entry_hash`) or `null` if empty. */
  head(): Hex64 | null;
  /** Number of entries written so far. */
  size(): number;
  /** Produce a manifest pinning the current state. */
  buildManifest(): JournalManifest;
}

// ─── MemoryJournal ─────────────────────────────────────────────────────

export class MemoryJournal implements Journal {
  private entries: SpinJournalEntry[] = [];
  private headHash: Hex64 | null = null;
  private nextSeq = 0;

  append(draft: Omit<SpinJournalEntry, 'prev_hash' | 'entry_hash'>): SpinJournalEntry {
    if (draft.seq !== this.nextSeq) {
      throw new Error(`MemoryJournal: expected seq ${this.nextSeq}, got ${draft.seq}`);
    }
    if (draft.schema_version !== RECALL_SCHEMA_VERSION) {
      throw new Error(
        `MemoryJournal: schema_version mismatch (want ${RECALL_SCHEMA_VERSION}, got ${draft.schema_version})`,
      );
    }
    const sealed = sealEntry(draft, this.headHash);
    this.entries.push(sealed);
    this.headHash = sealed.entry_hash;
    this.nextSeq++;
    return sealed;
  }

  readAll(): SpinJournalEntry[] {
    return this.entries.slice();
  }

  head(): Hex64 | null {
    return this.headHash;
  }

  size(): number {
    return this.entries.length;
  }

  buildManifest(): JournalManifest {
    return buildManifestFor(this.entries, '<memory>');
  }
}

// ─── NdjsonFileJournal ─────────────────────────────────────────────────

export class NdjsonFileJournal implements Journal {
  private headHash: Hex64 | null = null;
  private nextSeq = 0;
  private count = 0;
  private rotationMaxBytes: number;

  /**
   * @param path Path to the NDJSON file. If it exists, the journal
   *   recovers its head + seq counter from the last line.
   * @param rotationMaxBytes Rotate when the file crosses this size.
   *   The new file's first entry's `prev_hash` chains from the rotated
   *   one. Default 256 MiB.
   */
  constructor(
    private readonly path: string,
    rotationMaxBytes = 256 * 1024 * 1024,
  ) {
    this.rotationMaxBytes = rotationMaxBytes;
    this.recover();
  }

  private recover(): void {
    if (!existsSync(this.path)) return;
    const raw = readFileSync(this.path, 'utf8');
    if (raw.length === 0) return;
    const lines = raw.split('\n').filter((l) => l.length > 0);
    if (lines.length === 0) return;
    for (const line of lines) {
      const e = JSON.parse(line) as SpinJournalEntry;
      // Streaming verification — fail loud on tampered tail.
      if (e.prev_hash !== (this.headHash ?? ZERO_HASH)) {
        throw new Error(
          `NdjsonFileJournal.recover: chain break at seq ${e.seq} — expected prev_hash ${this.headHash ?? ZERO_HASH}, got ${e.prev_hash}`,
        );
      }
      if (e.seq !== this.nextSeq) {
        throw new Error(
          `NdjsonFileJournal.recover: seq gap at line — expected ${this.nextSeq}, got ${e.seq}`,
        );
      }
      this.headHash = e.entry_hash;
      this.nextSeq = e.seq + 1;
      this.count++;
    }
  }

  append(draft: Omit<SpinJournalEntry, 'prev_hash' | 'entry_hash'>): SpinJournalEntry {
    if (draft.seq !== this.nextSeq) {
      throw new Error(
        `NdjsonFileJournal: expected seq ${this.nextSeq}, got ${draft.seq}`,
      );
    }
    const sealed = sealEntry(draft, this.headHash);
    const line = JSON.stringify(sealed) + '\n';
    appendFileSync(this.path, line, 'utf8');
    this.headHash = sealed.entry_hash;
    this.nextSeq++;
    this.count++;
    this.maybeRotate();
    return sealed;
  }

  private maybeRotate(): void {
    try {
      const bytes = statSync(this.path).size;
      if (bytes < this.rotationMaxBytes) return;
    } catch {
      return;
    }
    const ts = new Date().toISOString().replace(/[:.]/g, '-');
    const rotatedPath = this.path.replace(/(\.ndjson)?$/, `-${ts}.ndjson`);
    renameSync(this.path, rotatedPath);
    // New file inherits the chain tail — first entry's `prev_hash`
    // will equal `this.headHash`, so the chain stays unbroken across
    // rotations. We leave `nextSeq` continuous so consumers walking
    // both files see a single monotonic sequence.
  }

  readAll(): SpinJournalEntry[] {
    if (!existsSync(this.path)) return [];
    const raw = readFileSync(this.path, 'utf8');
    if (raw.length === 0) return [];
    return raw
      .split('\n')
      .filter((l) => l.length > 0)
      .map((l) => JSON.parse(l) as SpinJournalEntry);
  }

  head(): Hex64 | null {
    return this.headHash;
  }

  size(): number {
    return this.count;
  }

  buildManifest(): JournalManifest {
    return buildManifestFor(this.readAll(), this.path);
  }
}

// ─── Helpers ────────────────────────────────────────────────────────────

function buildManifestFor(
  entries: SpinJournalEntry[],
  journalFile: string,
): JournalManifest {
  if (entries.length === 0) {
    return sealManifest({
      schema_version: RECALL_SCHEMA_VERSION as SchemaVersion,
      engine_version: getEngineVersion(),
      journal_file: journalFile,
      first_seq: 0,
      last_seq: -1,
      first_timestamp_utc: '',
      last_timestamp_utc: '',
      last_entry_hash: ZERO_HASH,
    });
  }
  // Verifying as we manifest is cheap and catches drift before a
  // regulator does.
  const ver = verifyChain(entries);
  if (!ver.ok) {
    throw new Error(`buildManifest: chain verification failed at seq=${ver.seq}: ${ver.detail}`);
  }
  const first = entries[0];
  const last = entries[entries.length - 1];
  return sealManifest({
    schema_version: RECALL_SCHEMA_VERSION as SchemaVersion,
    engine_version: getEngineVersion(),
    journal_file: journalFile,
    first_seq: first.seq,
    last_seq: last.seq,
    first_timestamp_utc: first.timestamp_utc,
    last_timestamp_utc: last.timestamp_utc,
    last_entry_hash: last.entry_hash,
  });
}

/**
 * Engine version surfaced into every manifest. Pulled from the
 * package.json at build time would be cleanest but the engine doesn't
 * ship a build step — keep this as a single source-of-truth constant.
 */
export function getEngineVersion(): string {
  // Keep in sync with rust-sim/Cargo.toml `version`.
  return '0.5.0';
}

/** Write a manifest atomically (write to .tmp + rename). */
export function writeManifest(path: string, manifest: JournalManifest): void {
  const tmp = path + '.tmp';
  writeFileSync(tmp, JSON.stringify(manifest, null, 2));
  renameSync(tmp, path);
}

/** Load a manifest, return `null` if file missing. */
export function readManifest(path: string): JournalManifest | null {
  if (!existsSync(path)) return null;
  return JSON.parse(readFileSync(path, 'utf8')) as JournalManifest;
}

File: rust-sim/src/recall/journal.rs
//! Memory + NDJSON file journals (Rust mirror).
//!
//! Same `Journal` interface as TS — `append`, `read_all`, `head`,
//! `size`, `build_manifest`. The file format is line-compatible: a
//! file written by TS can be read here and vice versa.

use super::integrity::{seal_entry, seal_manifest, verify_chain, ChainVerification};
use super::types::*;
use std::fs::{rename, File, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::Path;

/// Keep in sync with TS `src/recall/journal.ts::getEngineVersion()` and
/// `Cargo.toml` `version`.
pub const ENGINE_VERSION: &str = "0.5.0";

const DEFAULT_ROTATION_MAX_BYTES: u64 = 256 * 1024 * 1024;

pub trait Journal {
    /// Append a draft entry. The draft's `prev_hash` / `entry_hash`
    /// fields are ignored on input — the journal stamps them.
    fn append(&mut self, draft: SpinJournalEntry) -> Result<SpinJournalEntry, String>;
    fn read_all(&self) -> Vec<SpinJournalEntry>;
    fn head(&self) -> Option<String>;
    fn size(&self) -> usize;
    fn build_manifest(&self) -> JournalManifest;
}

// ─── MemoryJournal ──────────────────────────────────────────────────────

pub struct MemoryJournal {
    entries: Vec<SpinJournalEntry>,
    head_hash: Option<String>,
    next_seq: u64,
}

impl Default for MemoryJournal {
    fn default() -> Self {
        Self::new()
    }
}

impl MemoryJournal {
    pub fn new() -> Self {
        Self {
            entries: Vec::new(),
            head_hash: None,
            next_seq: 0,
        }
    }
}

impl Journal for MemoryJournal {
    fn append(&mut self, draft: SpinJournalEntry) -> Result<SpinJournalEntry, String> {
        if draft.seq != self.next_seq {
            return Err(format!(
                "MemoryJournal: expected seq {}, got {}",
                self.next_seq, draft.seq
            ));
        }
        if draft.schema_version != RECALL_SCHEMA_VERSION {
            return Err(format!(
                "MemoryJournal: schema_version mismatch (want {}, got {})",
                RECALL_SCHEMA_VERSION, draft.schema_version
            ));
        }
        let sealed = seal_entry(draft, self.head_hash.as_deref());
        self.head_hash = Some(sealed.entry_hash.clone());
        self.next_seq += 1;
        self.entries.push(sealed.clone());
        Ok(sealed)
    }
    fn read_all(&self) -> Vec<SpinJournalEntry> {
        self.entries.clone()
    }
    fn head(&self) -> Option<String> {
        self.head_hash.clone()
    }
    fn size(&self) -> usize {
        self.entries.len()
    }
    fn build_manifest(&self) -> JournalManifest {
        build_manifest_for(&self.entries, "<memory>")
    }
}

// ─── NdjsonFileJournal ──────────────────────────────────────────────────

pub struct NdjsonFileJournal {
    path: String,
    head_hash: Option<String>,
    next_seq: u64,
    count: usize,
    rotation_max_bytes: u64,
}

impl NdjsonFileJournal {
    pub fn new(path: impl Into<String>) -> Result<Self, String> {
        Self::with_rotation(path, DEFAULT_ROTATION_MAX_BYTES)
    }

    pub fn with_rotation(path: impl Into<String>, rotation_max_bytes: u64) -> Result<Self, String> {
        let mut j = Self {
            path: path.into(),
            head_hash: None,
            next_seq: 0,
            count: 0,
            rotation_max_bytes,
        };
        j.recover()?;
        Ok(j)
    }

    fn recover(&mut self) -> Result<(), String> {
        let path = Path::new(&self.path);
        if !path.exists() {
            return Ok(());
        }
        let file = File::open(path).map_err(|e| format!("open {}: {e}", self.path))?;
        let reader = BufReader::new(file);
        for (line_no, line) in reader.lines().enumerate() {
            let line = line.map_err(|e| format!("read line {line_no}: {e}"))?;
            if line.is_empty() {
                continue;
            }
            let e: SpinJournalEntry = serde_json::from_str(&line)
                .map_err(|err| format!("parse line {line_no}: {err}"))?;
            let want_prev = self
                .head_hash
                .clone()
                .unwrap_or_else(|| ZERO_HASH.to_string());
            if e.prev_hash != want_prev {
                return Err(format!(
                    "NdjsonFileJournal::recover: chain break at seq {} — expected prev_hash {}, got {}",
                    e.seq, want_prev, e.prev_hash
                ));
            }
            if e.seq != self.next_seq {
                return Err(format!(
                    "NdjsonFileJournal::recover: seq gap at line {line_no} — expected {}, got {}",
                    self.next_seq, e.seq
                ));
            }
            self.head_hash = Some(e.entry_hash.clone());
            self.next_seq = e.seq + 1;
            self.count += 1;
        }
        Ok(())
    }

    fn maybe_rotate(&self) {
        let path = Path::new(&self.path);
        let Ok(meta) = std::fs::metadata(path) else {
            return;
        };
        if meta.len() < self.rotation_max_bytes {
            return;
        }
        let ts = chrono_like_utc();
        let rotated = if let Some(stripped) = self.path.strip_suffix(".ndjson") {
            format!("{stripped}-{ts}.ndjson")
        } else {
            format!("{}-{ts}.ndjson", self.path)
        };
        let _ = rename(&self.path, &rotated);
    }
}

impl Journal for NdjsonFileJournal {
    fn append(&mut self, draft: SpinJournalEntry) -> Result<SpinJournalEntry, String> {
        if draft.seq != self.next_seq {
            return Err(format!(
                "NdjsonFileJournal: expected seq {}, got {}",
                self.next_seq, draft.seq
            ));
        }
        if draft.schema_version != RECALL_SCHEMA_VERSION {
            return Err(format!(
                "NdjsonFileJournal: schema_version mismatch (want {}, got {})",
                RECALL_SCHEMA_VERSION, draft.schema_version
            ));
        }
        let sealed = seal_entry(draft, self.head_hash.as_deref());
        // serde_json's default emitter doesn't pretty-print and emits
        // valid JSON for each line — fine for NDJSON.
        let line = serde_json::to_string(&sealed)
            .map_err(|e| format!("NdjsonFileJournal: serialize: {e}"))?;
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)
            .map_err(|e| format!("open {} for append: {e}", self.path))?;
        writeln!(file, "{}", line).map_err(|e| format!("write line: {e}"))?;
        drop(file);
        self.head_hash = Some(sealed.entry_hash.clone());
        self.next_seq += 1;
        self.count += 1;
        self.maybe_rotate();
        Ok(sealed)
    }
    fn read_all(&self) -> Vec<SpinJournalEntry> {
        let path = Path::new(&self.path);
        if !path.exists() {
            return Vec::new();
        }
        let file = match File::open(path) {
            Ok(f) => f,
            Err(_) => return Vec::new(),
        };
        let reader = BufReader::new(file);
        let mut out = Vec::new();
        for line in reader.lines().map_while(Result::ok) {
            if line.is_empty() {
                continue;
            }
            if let Ok(e) = serde_json::from_str::<SpinJournalEntry>(&line) {
                out.push(e);
            }
        }
        out
    }
    fn head(&self) -> Option<String> {
        self.head_hash.clone()
    }
    fn size(&self) -> usize {
        self.count
    }
    fn build_manifest(&self) -> JournalManifest {
        build_manifest_for(&self.read_all(), &self.path)
    }
}

fn build_manifest_for(entries: &[SpinJournalEntry], journal_file: &str) -> JournalManifest {
    if entries.is_empty() {
        return seal_manifest(JournalManifest {
            schema_version: RECALL_SCHEMA_VERSION.into(),
            engine_version: ENGINE_VERSION.into(),
            journal_file: journal_file.into(),
            first_seq: 0,
            last_seq: -1,
            first_timestamp_utc: String::new(),
            last_timestamp_utc: String::new(),
            last_entry_hash: ZERO_HASH.into(),
            manifest_hash: String::new(),
        });
    }
    match verify_chain(entries) {
        ChainVerification::Ok { .. } => {}
        other => panic!("build_manifest: chain verification failed: {other:?}"),
    }
    let first = &entries[0];
    let last = &entries[entries.len() - 1];
    seal_manifest(JournalManifest {
        schema_version: RECALL_SCHEMA_VERSION.into(),
        engine_version: ENGINE_VERSION.into(),
        journal_file: journal_file.into(),
        first_seq: first.seq as i64,
        last_seq: last.seq as i64,
        first_timestamp_utc: first.timestamp_utc.clone(),
        last_timestamp_utc: last.timestamp_utc.clone(),
        last_entry_hash: last.entry_hash.clone(),
        manifest_hash: String::new(),
    })
}

pub fn write_manifest(path: &str, manifest: &JournalManifest) -> Result<(), String> {
    let tmp = format!("{path}.tmp");
    let payload = serde_json::to_string_pretty(manifest).map_err(|e| format!("serialize: {e}"))?;
    let mut file = File::create(&tmp).map_err(|e| format!("create {tmp}: {e}"))?;
    file.write_all(payload.as_bytes())
        .map_err(|e| format!("write: {e}"))?;
    file.sync_all().map_err(|e| format!("sync: {e}"))?;
    rename(&tmp, path).map_err(|e| format!("rename {tmp} → {path}: {e}"))?;
    Ok(())
}

pub fn read_manifest(path: &str) -> Result<Option<JournalManifest>, String> {
    let p = Path::new(path);
    if !p.exists() {
        return Ok(None);
    }
    let raw = std::fs::read_to_string(p).map_err(|e| format!("read {path}: {e}"))?;
    let m: JournalManifest =
        serde_json::from_str(&raw).map_err(|e| format!("parse {path}: {e}"))?;
    Ok(Some(m))
}

/// Cheap ISO8601-ish UTC timestamp for file rotation. Avoids pulling
/// `chrono` for one filename suffix.
fn chrono_like_utc() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    // Suffix is monotonic and unique-enough — full calendar conversion
    // is overkill since this is filesystem metadata, not journal content.
    format!("epoch-{secs}")
}
